_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+
+```json
+{
+    "error": "invalid_grant",
+    "error_description": "PKCE code verifier not specified"
+}
+```
+
+This is why the code flow + PKCE is more secure than the implicit flow. 
+Even if an attacker manages to obtain the authorization grant, it's worthless without the `code_verifier`.
+
+> Note that the HTTP 400 will only occur when using PKCE. 
+If no PKCE is used, the client should be confidential (requiring credentials to exchange the authorization grant) rather than be public.
+
+So only our Angular client will be able to retrieve the access token in the form of a JSON Web Token.
+
+## JSON Web Token (JWT)
+JSON Web Tokens or JWT, often pronounced as 'jot', is an open standard for a compact way of representing claims to be transferred between two parties. 
+What this means is that it's a special kind of object which has some data in it. It can also be digitally signed to make sure it is not tampered with. 
+
+In its compact form, JSON Web Tokens consist of three parts separated by dots (.), which are:
+
+* `Header`:  the type of the token and the signing algorithm being used
+* `Payload`: the payload, which contains the claims and additional data
+* `Signature`: to verify if the token was not tampered with
+
+Therefore, a JWT typically looks like the following:
+
+```
+xxxxx.yyyyy.zzzzz
+```
+
+The header (xxx) and payload (yyy) are base64 encoded. 
+An access token is a good example of a JWT:
+
+```
+eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJkdE9JZkY2NGZRYnlwWVRFdGV2eV83NUdIWTdQMmNHU1o2a2ZXWDdFblBJIn0.eyJqdGkiOiIxY2EzZTZkYS1kM2Y2LTRiYTMtYjNjZC1iMDExYmRlM2JmNmIiLCJleHAiOjE1NjYzMjk1NTYsIm5iZiI6MCwiaWF0IjoxNTY2MzI5MjU2LCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAvYXV0aC9yZWFsbXMvaGVyb2VzIiwiYXVkIjoiYWNjb3VudCIsInN1YiI6IjBkNjg0OWI4LWUyZmEtNGU3My04NjlhLTE1ZDVhOTE1YzdhMiIsInR5cCI6IkJlYXJlciIsImF6cCI6InNwYS1oZXJvZXMiLCJhdXRoX3RpbWUiOjAsInNlc3Npb25fc3RhdGUiOiI4NWRjYTg0Ny00YmQzLTRiOTUtOTNiYy01MjE5ZjUzYWNiMzciLCJhY3IiOiIxIiwiYWxsb3dlZC1vcmlnaW5zIjpbIioiXSwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbIm9mZmxpbmVfYWNjZXNzIiwidW1hX2F1dGhvcml6YXRpb24iLCJ1c2VyIl19LCJyZXNvdXJjZV9hY2Nlc3MiOnsiYWNjb3VudCI6eyJyb2xlcyI6WyJtYW5hZ2UtYWNjb3VudCIsIm1hbmFnZS1hY2NvdW50LWxpbmtzIiwidmlldy1wcm9maWxlIl19fSwic2NvcGUiOiJlbWFpbCBoZXJvZXMgcHJvZmlsZSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJuYW1lIjoiSmVyb2VuIE1leXMiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJqZXJvZW4iLCJnaXZlbl9uYW1lIjoiSmVyb2VuIiwiZmFtaWx5X25hbWUiOiJNZXlzIiwiZW1haWwiOiJtZUBhY21lLmJlIn0.cvn79d-n0aFYqDB3p-htDNeeYOdkvqEsmDhGKp9V3a4i6nJx7dU0_r7zicQe26ZgDsM65ILx_X-buWv-e5_eraFo1OOveCGtBbrrLwrQ0Z7SlVMHJrDooJrLEE_m8Qlz_-iLcEC2-ODroEwyLRej_Du626B48QL2bcq-8ADqGSaLf7Y4ZTVMiP_p6dsCi4GDQLq1WOy-g6--z47FKTJVuAl2yY_JNNuEd5aofw0FTE38EoEinIdcy5NXCXDhtGHr_k5lA2Swu4JvK84YB6usECigCb1_zO_c6LhZQkRTCcCojxC6Qn1trQH9epcFEKTkDCHrNf6BLp4X9rH2URWJcA
+```
+
+We can easily decode them using online tools like [jwt.io](https://jwt.io/).
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/jwt.png" width="70%">
```

Comment:
Add an image with eg. `<img class="image fit" src="{{ '/img/devoxx-ma-2018/speakers-dinner.jpg' | prepend: site.baseurl }}" alt="Speakers Dinner" width="60%" />`, otherwise this won't scale good.